File: tests/algorithms/x/PHP/ciphers/xor_cipher.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
function _intdiv($a, $b) {
    if (function_exists('bcdiv')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcdiv($sa, $sb, 0));
    }
    return intdiv($a, $b);
}
$__start_mem = memory_get_usage();
$__start = _now();
  function mochi_xor($a, $b) {
  global $ascii, $sample, $enc, $dec;
  $res = 0;
  $bit = 1;
  $x = $a;
  $y = $b;
  while ($x > 0 || $y > 0) {
  $abit = $x % 2;
  $bbit = $y % 2;
  if ($abit != $bbit) {
  $res = $res + $bit;
}
  $x = _intdiv($x, 2);
  $y = _intdiv($y, 2);
  $bit = $bit * 2;
};
  return $res;
};
  $ascii = ' !"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~';
  function mochi_ord($ch) {
  global $ascii, $sample, $enc, $dec;
  $i = 0;
  while ($i < strlen($ascii)) {
  if (substr($ascii, $i, $i + 1 - $i) == $ch) {
  return 32 + $i;
}
  $i = $i + 1;
};
  return 0;
};
  function mochi_chr($n) {
  global $ascii, $sample, $enc, $dec;
  if ($n >= 32 && $n < 127) {
  return substr($ascii, $n - 32, $n - 31 - ($n - 32));
}
  return '';
};
  function normalize_key($key) {
  global $ascii, $sample, $enc, $dec;
  $k = $key;
  if ($k == 0) {
  $k = 1;
}
  $k = $k % 256;
  if ($k < 0) {
  $k = $k + 256;
}
  return $k;
};
  function encrypt($content, $key) {
  global $ascii, $sample, $enc, $dec;
  $k = normalize_key($key);
  $result = [];
  $i = 0;
  while ($i < strlen($content)) {
  $c = mochi_ord(substr($content, $i, $i + 1 - $i));
  $e = mochi_xor($c, $k);
  $result = _append($result, mochi_chr($e));
  $i = $i + 1;
};
  return $result;
};
  function encrypt_string($content, $key) {
  global $ascii, $sample, $enc, $dec;
  $chars = encrypt($content, $key);
  $out = '';
  foreach ($chars as $ch) {
  $out = $out . $ch;
};
  return $out;
};
  $sample = 'hallo welt';
  $enc = encrypt_string($sample, 1);
  $dec = encrypt_string($enc, 1);
  echo rtrim(_str(encrypt($sample, 1))), PHP_EOL;
  echo rtrim($enc), PHP_EOL;
  echo rtrim($dec), PHP_EOL;
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;

File: tests/algorithms/transpiler/FS/ciphers/xor_cipher.fs
// Generated 2025-08-07 10:31 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let rec _str v =
    let s = sprintf "%A" v
    s.Replace("[|", "[")
     .Replace("|]", "]")
     .Replace("; ", " ")
     .Replace(";", "")
     .Replace("\"", "")
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let rec xor (a: int) (b: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable a = a
    let mutable b = b
    try
        let mutable res: int = 0
        let mutable bit: int = 1
        let mutable x: int = a
        let mutable y: int = b
        while (x > 0) || (y > 0) do
            let abit: int = ((x % 2 + 2) % 2)
            let bbit: int = ((y % 2 + 2) % 2)
            if abit <> bbit then
                res <- res + bit
            x <- x / 2
            y <- y / 2
            bit <- bit * 2
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let ascii: string = " !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~"
let rec ord (ch: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable ch = ch
    try
        let mutable i: int = 0
        while i < (String.length (ascii)) do
            if (ascii.Substring(i, (i + 1) - i)) = ch then
                __ret <- 32 + i
                raise Return
            i <- i + 1
        __ret <- 0
        raise Return
        __ret
    with
        | Return -> __ret
let rec chr (n: int) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable n = n
    try
        __ret <- if (n >= 32) && (n < 127) then (ascii.Substring(n - 32, (n - 31) - (n - 32))) else ""
        raise Return
        __ret
    with
        | Return -> __ret
let rec normalize_key (key: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable key = key
    try
        let mutable k: int = key
        if k = 0 then
            k <- 1
        k <- ((k % 256 + 256) % 256)
        if k < 0 then
            k <- k + 256
        __ret <- k
        raise Return
        __ret
    with
        | Return -> __ret
let rec encrypt (content: string) (key: int) =
    let mutable __ret : string array = Unchecked.defaultof<string array>
    let mutable content = content
    let mutable key = key
    try
        let mutable k: int = normalize_key (key)
        let mutable result: string array = [||]
        let mutable i: int = 0
        while i < (String.length (content)) do
            let c: int = ord (content.Substring(i, (i + 1) - i))
            let e: int = xor (c) (k)
            result <- Array.append result [|chr (e)|]
            i <- i + 1
        __ret <- result
        raise Return
        __ret
    with
        | Return -> __ret
let rec encrypt_string (content: string) (key: int) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable content = content
    let mutable key = key
    try
        let chars: string array = encrypt (content) (key)
        let mutable out: string = ""
        for ch in Seq.map string (chars) do
            out <- out + ch
        __ret <- out
        raise Return
        __ret
    with
        | Return -> __ret
let sample: string = "hallo welt"
let enc: string = encrypt_string (sample) (1)
let dec: string = encrypt_string (enc) (1)
printfn "%s" (_str (encrypt (sample) (1)))
printfn "%s" (enc)
printfn "%s" (dec)
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)
